### ISSUE TITLE
Worker timeout

### ISSUE THREAD

**Issue by nicofrand** [caronc/apprise-api#16]
:question: **Question**

Hi! 

Thank you very much for both the project and API! I am using the ![docker image](https://hub.docker.com/r/caronc/apprise) to post request to Apprise as a gateway to Gotify.

It works most of the time but at night I often miss some notifications. There are some logs but I don't understand them:
what triggers this worker timeout? The app calling apprise-api, or the gotify server that apprise-api calls?

```
[2020-07-28 22:42:27 +0000] [183] [CRITICAL] WORKER TIMEOUT (pid:201)
2020/07/28 22:42:31 [error] 198#198: *31 upstream prematurely closed connection while reading response header from upstream, client: 172.17.0.1, server: _, request: "POST /notify HTTP/1.1", upstream: "http://127.0.0.1:8080/notify", host: "localhost:8000"
2020/07/28 22:42:31 [warn] 198#198: *31 upstream server temporarily disabled while reading response header from upstream, client: 172.17.0.1, server: _, request: "POST /notify HTTP/1.1", upstream: "http://127.0.0.1:8080/notify", host: "localhost:8000"
172.17.0.1 - - [28/Jul/2020:22:42:32 +0000] "POST /notify HTTP/1.1" 502 494 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
[2020-07-28 22:42:33 +0000] [210] [INFO] Booting worker with pid: 210
```
Logs since the start if needed (all the POST requests are coming from successful tests):

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.                                                                                                                                     
[s6-init] ensuring user provided files have correct perms...exited 0.                                                                                                                                             
[fix-attrs.d] applying ownership & permissions fixes...                                                                                                                                                           
[fix-attrs.d] done.                                                                                                                                                                                               
[cont-init.d] executing container initialization scripts...                                                                                                                                                       
[cont-init.d] done.                                                                                                                                                                                               
[services.d] starting services                                                                                                                                                                                    
[services.d] done.                                                                                                                                                                                                
Starting gunicorn                                                                                                                                                                                                 
Starting nginx                                                                                                                                                                                                    
2020/07/28 17:53:57 [notice] 197#197: using the "epoll" event method                                                                                                                                              
2020/07/28 17:53:57 [notice] 197#197: nginx/1.19.0
2020/07/28 17:53:57 [notice] 197#197: built by gcc 8.3.0 (Debian 8.3.0-6)
2020/07/28 17:53:57 [notice] 197#197: OS: Linux 4.9.0-12-amd64
2020/07/28 17:53:57 [notice] 197#197: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2020/07/28 17:53:57 [notice] 197#197: start worker processes
2020/07/28 17:53:57 [notice] 197#197: start worker process 198
[2020-07-28 17:53:58 +0000] [183] [INFO] Starting gunicorn 20.0.4
[2020-07-28 17:53:58 +0000] [183] [INFO] Listening at: http://0.0.0.0:8080 (183)
[2020-07-28 17:53:58 +0000] [183] [INFO] Using worker: sync
[2020-07-28 17:53:58 +0000] [201] [INFO] Booting worker with pid: 201
[2020-07-28 17:53:58 +0000] [202] [INFO] Booting worker with pid: 202
[2020-07-28 17:53:58 +0000] [203] [INFO] Booting worker with pid: 203
[2020-07-28 17:53:58 +0000] [204] [INFO] Booting worker with pid: 204
[2020-07-28 17:53:58 +0000] [205] [INFO] Booting worker with pid: 205
[2020-07-28 17:53:58 +0000] [206] [INFO] Booting worker with pid: 206
[2020-07-28 17:53:58 +0000] [207] [INFO] Booting worker with pid: 207
[2020-07-28 17:53:58 +0000] [208] [INFO] Booting worker with pid: 208
[2020-07-28 17:53:59 +0000] [209] [INFO] Booting worker with pid: 209
127.0.0.1 - - [28/Jul/2020:13:00:57 -0500] "GET / HTTP/1.0" 200 25873 "-" "curl/7.52.1"
172.17.0.1 - - [28/Jul/2020:18:00:57 +0000] "GET / HTTP/1.1" 200 25873 "-" "curl/7.52.1" "-"
2020/07/28 18:00:57 [info] 198#198: *1 client 172.17.0.1 closed keepalive connection
127.0.0.1 - - [28/Jul/2020:13:01:05 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:01:06 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:01:14 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:01:14 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:02:09 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:02:09 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:02:12 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:02:12 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:04:52 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:04:52 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:04:57 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:04:57 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:04:59 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:04:59 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:05:11 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:05:11 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:05:12 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:05:12 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:05:13 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [28/Jul/2020:18:05:13 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [28/Jul/2020:13:05:15 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
```

Can you help please? Thanks in advance!

**Comment by caronc** [comment]
Hi,

I think your problem ties down to a recent change i made [here](https://github.com/caronc/apprise/issues/244) where a timeout was introduced.  The timeout's i chose seem a bit too aggressive. [This is the section in question](https://github.com/caronc/apprise/blob/57cb5f38d214116d45d7631d2a3c5e7e9251516b/apprise/plugins/NotifyGotify.py#L83-L86).

Perhaps 2.5 seconds read timeout should be bumped to say... 4.0... or maybe 5.0 (doubling it)?
* Too low and we get symptoms like you're seeing.
* Too high and response times can be slow for some systems/users/services.

It's trying to find that sweet spot... :slightly_smiling_face: 

Globally (all of Apprise) uses [the same timeouts](https://github.com/caronc/apprise/blob/57cb5f38d214116d45d7631d2a3c5e7e9251516b/apprise/URLBase.py#L101-L108).

I'll created [this ticket here](https://github.com/caronc/apprise/issues/262) to help track this further since the fix won't end up being part of the _Apprise API_.

Thank you for reporting this issue!

Chris

**Comment by nicofrand** [comment]
Yeah, my server is quite old but I agree 2.5 seconds to do a request (on the same machine, just outside the docker) is way too much :o.

Maybe a solution would be to allow to configure the timeout, so I can increase it myself without breaking the feature for other users?

**Comment by caronc** [comment]
> Maybe a solution would be to allow to configure the timeout, so I can increase it myself without breaking the feature for other users?

I love this idea; i think what I'll do is just bump the 2.5 seconds to **4.0** seconds (just a wee bit more then it is and may fix your issue right out of the box).

But (taking your advice) I'll also add the variables `read_timeout=` and `connect_timeout` accessible from **all** Apprise URLs that deal with web/api query based services too (for consistency). 

You can over-ride both and/or just one of them; so basically your Apprise URL will look like:
* `gotify://url/details/here?read_timeout=15` (_as an example_).
* `gotify://url/details/here?connect_timeout=15` (_as an example_).
* `gotify://url/details/here?connect_timeout=15&read_timeout=30` (_as an example_).

I'll hopefully have this fixed before the end of the weekend (Aug 2nd); I'll keep you posted here (as changes will be mostly reflected in the other ticket where the code base actually resides).

**Comment by nicofrand** [comment]
Wow that would be really great. I don't control the calls from our app (Kresus) to apprise (only the base URL) but I do control the URL to gotify so that would work perfectly for me!

Take your time, no stress.

**Comment by caronc** [comment]
I updated my other comment (above) to reflect the changes made to Apprise in the pull request.  I also triggered a new build on dockerhub, so you should be able to pull in these changes now and try it out.

Let me know how it goes! :slightly_smiling_face:

**Comment by nicofrand** [comment]
Thank you very much!
For now I updated my install and will monitor it with the 4 seconds timeout. If that's still not enough I will use the  rto and cto parameters.

Thanks again for the answers and quick "fix"!

**Comment by caronc** [comment]
I don't think 4 seconds is that short of a time frame at all; here is how it is documented (from the [source](https://2.python-requests.org/en/master/user/advanced/#timeouts)):

> the read timeout is the number of seconds the client will wait for the server to send a response. (**Specifically, it’s the number of seconds that the client will wait between bytes sent from the server**. In 99.9% of cases, this is the time before the server sends the first byte).

I bolded the important part. It's not how long the entire transmission between Apprise and your Gotify server.  It's how long we continue waiting for data (Gotify is still holding the connection open) when there is absolutely no data being exchanged at all.  If it takes 15 seconds to communicate with your Gotify Server, but it's still actively sending content then the connection will remain open.  4 seconds is a very, very long stall time (especially for computers, even slow ones).

I'll leave this ticket open for a few more days, but I'm definitely interested to know if we fix your problem, or if we need to keep digging. :slightly_smiling_face:

**Comment by nicofrand** [comment]
It does not seem enough to fix my issue, but I don't know in which state my server is around those hours. Maybe all the resources are dedicated to another task like a backup, resulting in a timeout from my Gotify…

```
2020/08/04 03:48:33 [error] 198#198: *13 no live upstreams while connecting to upstream, client: 172.17.0.1, server: _, request: "POST /notify HTTP/1.1", upstream: "http://apprise_backend/notify", host: "localhost:8000"
172.17.0.1 - - [04/Aug/2020:03:48:33 +0000] "POST /notify HTTP/1.1" 502 494 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
127.0.0.1 - - [03/Aug/2020:22:48:54 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [04/Aug/2020:03:48:55 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
[2020-08-04 03:49:04 +0000] [183] [CRITICAL] WORKER TIMEOUT (pid:204)
2020/08/04 03:49:05 [error] 198#198: *9 upstream prematurely closed connection while reading response header from upstream, client: 172.17.0.1, server: _, request: "POST /notify HTTP/1.1", upstream: "http://127.0.0.1:8080/notify", host: "localhost:8000"
2020/08/04 03:49:05 [warn] 198#198: *9 upstream server temporarily disabled while reading response header from upstream, client: 172.17.0.1, server: _, request: "POST /notify HTTP/1.1", upstream: "http://127.0.0.1:8080/notify", host: "localhost:8000"
172.17.0.1 - - [04/Aug/2020:03:49:05 +0000] "POST /notify HTTP/1.1" 502 494 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
[2020-08-04 03:49:05 +0000] [212] [INFO] Booting worker with pid: 212
```

**Comment by caronc** [comment]
Well, the good news is you have control over the timeout now (I'm not keen on increasing it's value any more).  So you could try setting something like `?rto=60&cto=15` on your URL that you load into the Apprise-API

The other thing I could try to implement is a `retry=` where it would default to zero (0).  But if you set it to say... 1,2, or 3 it would retry the notification that many more times before all out giving up and returning a failure.

**Comment by nicofrand** [comment]
Just tried with `?priority=high&rto=180&cto=45`

and I still have issues:

```
127.0.0.1 - - [04/Aug/2020:09:50:00 -0500] "POST /notify HTTP/1.0" 200 21 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
172.17.0.1 - - [04/Aug/2020:14:50:00 +0000] "POST /notify HTTP/1.1" 200 52 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
[2020-08-04 14:50:33 +0000] [183] [CRITICAL] WORKER TIMEOUT (pid:211)
2020/08/04 14:50:34 [error] 198#198: *62 upstream prematurely closed connection while reading response header from upstream, client: 172.17.0.1, server: _, request: "POST /notify HTTP/1.1", upstream: "http://127.0.0.1:8080/notify", host: "localhost:8000"
2020/08/04 14:50:34 [warn] 198#198: *62 upstream server temporarily disabled while reading response header from upstream, client: 172.17.0.1, server: _, request: "POST /notify HTTP/1.1", upstream: "http://127.0.0.1:8080/notify", host: "localhost:8000"
172.17.0.1 - - [04/Aug/2020:14:50:35 +0000] "POST /notify HTTP/1.1" 502 494 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
[2020-08-04 14:50:35 +0000] [217] [INFO] Booting worker with pid: 217
```

So I guess the issue comes from something else on my server (I checked iptables/fail2ban) but I still have no clue what the cause is…

Anyway, thank you very much for those parameters, I am sure I'll need them sometimes anyway!

**Comment by caronc** [comment]
According to [Gunicorn Docs](https://docs.gunicorn.org/en/stable/settings.html#timeout) it will abort the connection after 30 seconds.  This Pull request triples up on that.  It may help you out.  To test and see if it works for you, get your docker container:
```bash
# Find the container running gunicorn:
docker ps | egrep -i apprise | egrep gunicorn
# output might look like:
# 49d3b4732d19        apprise-api_backend    "gunicorn -c /opt/ap…"   ....
#        ^
#        |
#  container name
#
Get inside of the container: 
# docker exec -it <container name> /bin/sh

# Update timeout:
echo "timeout = 90" >> /opt/apprise/webapp/gunicorn.conf.py

# exit from within docker container
exit

# Now restart the container (so it reads the new configuration)
docker restart <container name>
```

If that works for you, I'll merge the branch I just created.  Sound good?

**Comment by caronc** [comment]
HI i hadn't heard back from you in a bit.  I was wondering if i could close this issue off? Did you end up figuring it out?

**Comment by nicofrand** [comment]
Hi, sorry I had some issues with my server then for the last days I did not have any issue sending the notifications, by chance, so I was waiting for it to come back to be sure the fix would actually work.

I increased the timeout to 90 like you said and will now try it more extensively (I have more time today), thanks!

**Comment by nicofrand** [comment]
Argh, it looks like I still issues with my server, I can't send any notification at all anymore (even without adding the timeout line).

From my client:

```
[2020-08-17T12:21:47.563] [ERROR] HTTP - POST /api/settings/test-notification/ - 424 (21440 ms)
```

In the apprise container logs (even after setting the timeout to 90):

```
127.0.0.1 - - [17/Aug/2020:07:21:47 -0500] "POST /notify HTTP/1.0" 424 43 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"                                                                                                      
172.17.0.1 - - [17/Aug/2020:12:21:47 +0000] "POST /notify HTTP/1.1" 424 43 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "-"
```

Feel free to close the issue anyway, I'll play with the timeout once I manage to make everything work again.

Sorry for the delay!

**Comment by caronc** [comment]
I'm sorry you're having all this trouble. I don't mind leaving the ticket open to assist in any way i can, but based on all of your issues, it does appear to be on your end  (_like you've already identified yourself_). :slightly_frowning_face: 

But now that I know you're still around, let's just leave the ticket open for a while longer.  I don't mind helping out in any way I can should there be something you find I might be able to help with!

Good luck!

**Comment by nicofrand** [comment]
I tried sending requests from my phone gotify's app and I can see it on my web app so that point should be ok.

I am now trying to go step by step and removing one piece (my client) by replacing it with a simpe curl request:

```
curl -X POST -d '{"urls": "gotifys://push.nicofrand.eu/MYTOKEN?priority=high&rto=60&cto=15", "body":"test message"}' -H "Content-Type: application/json" http://localhost:8000/notify
```

but it returns a `HTTP/1.1 424 Failed Dependency`. Could you tell if my request is correct?

**Comment by caronc** [comment]
424 maps to `One or more notifications could not be sent`.  It's tough to say at this point if it's because the token specified is bad or what.  If you use the the `apprise` tool on it's own (outside of the API), how does it work for you?

```bash
# Create a virtual environment to work in
# This way you can just destroy it after when it's all over.
# The below will create a directory called apprise
python3 -m venv apprise

# Change into our new directory
cd apprise

# Activate our virtual environment
source bin/activate

# Fetch the master for now to work with
pip install git+https://github.com/caronc/apprise.git

# Now try your URL:
# -D : Debug Mode (this is new to this master branch build)
# -b: Body
apprise -D -b "test message" \
   "gotifys://push.nicofrand.eu/MYTOKEN?priority=high&rto=60&cto=15"
```

**Comment by nicofrand** [comment]
Thanks!

Outside the API (and thus the docker container), it works perfectly. It must have to do with my docker config, sorry for the noise!
I'll be away for a week starting tomorrow and I don't think I'll have it fixed by then to test the timeout config, so feel free to close the issue!

**Comment by caronc** [comment]
You're honestly not wasting my time; if i have a bug, i definitely want to fix it.  I also don't mind trying to help you out if you find the issue could be on my end :slightly_smiling_face: 

I'll close this issue for now, but feel free to re-open it and/or open a new ticket whenever you like! I wish you the best of luck!

**Comment by nicofrand** [comment]
Hi again,

I don't have the issue right now after several tests (and a docker network change). Just wanted to thank you again:

The latest issue on my side were due to docker not having external network access after a debian upgrade and nftables migration. I am not savvy enough to fix it properly so I passed the `--network host` parameter but hit another issue: I already have something on port 8080. So I was about to build a new docker without nginx and only gunicorn. And then I saw you already provide such an image :heart:.
